### PR TITLE
[Clang](NFC) Add coverage for VTable debug info

### PR DESCRIPTION
**Pull request by kikairoya**
Add test for:
- shows effect of inlining member functions
- shows effect of template instantiation methods

Make cleaner a bit:
- drops unnecessary `REQUIRES` clause
- uses `%clang_cc1` instead of `%clang -Xclang`
- uses `-O0 -disable-llvm-passes` and `-O1 -disable-llvm-passes` because the optimization level affects the linkage of VTables.

### CODE REVIEW

**Comment by kikairoya** [comment]
I'd like to add test coverage and verify the intended behavior on edge cases before submitting the long-term fix for #149639.

@jmorse @CarlosAlbertoEnciso  
Could you take a look at this and check if the new tests reflect the intended behavior?

**Comment by CarlosAlbertoEnciso** [comment]
It seems that there are conflicts that must be resolved.

**Review by CarlosAlbertoEnciso — COMMENTED** [review]
Are you removing the `REQUIRES` because https://github.com/llvm/llvm-project/pull/150938 (platforms that don't generate DWARF)?

**Comment by kikairoya** [comment]
Thank you for looking this.

> Are you removing the `REQUIRES` because #150938 (platforms that don't generate DWARF)?

I believe these tests should be run in any configuration because they can be run. And, yes, we'll need to cover COFF cases, which will be added in a next PR.

**Comment by CarlosAlbertoEnciso** [comment]
The tests look good.

File: clang/test/DebugInfo/CXX/vtable-template-instantiation.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+//   * Its '_vtable$' is generated
+// - Define explicitly instantiation:
+//   * The vtable is generated with comdat
+//   * Its '_vtable$' is generated
+// - Declare explicitly instantiation as extern:
+//  # for COFF targets:
+//   * The vtable is declared but NOT associated with '_vtable$'
+//  # for non-COFF targets:
+//   * The vtable is declared
+//   * Its '_vtable$' is generated
```
Comment (jmorse) [inline clang/test/DebugInfo/CXX/vtable-template-instantiation.cpp:16]: To confirm my understanding -- what's interesting here is that the _vtable$ variable is generated in various scenarios where the vtable itself isn't generated -- thus the test is checking that _vtable$ references are generated for an external symbol?

**Review by jmorse — COMMENTED** [review]
I think adding / adjusting this coverage is good -- a few questions / requests inline.

File: clang/test/DebugInfo/CXX/vtable-external.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
+// The debug info of vtable is attached conditionally to whether
+// - Member functions are inlined or not
+// - Definition of destructor is visible or not
```
[the remark this answers is not in the document]
Comment (kikairoya) [reply clang/test/DebugInfo/CXX/vtable-external.cpp:3]: Sure. I'll try it.

File: clang/test/DebugInfo/CXX/vtable-inheritance-simple-main.cpp [hunk 3]
```diff
@@ -1,5 +1,3 @@
-// REQUIRES: target={{x86_64.*-linux.*}}
```
[the remark this answers is not in the document]
Comment (kikairoya) [reply clang/test/DebugInfo/CXX/vtable-inheritance-simple-main.cpp:1]: `%clang_cc1 -emit-llvm` for any targets should functional completely regardless which targets LLVM has been configured for.
Here is the test already working `-triple x86_64-linux` without REQUIRES clause.
https://github.com/llvm/llvm-project/blob/917f078adb3ffab4cf50aec3e13c977c092d7754/clang/test/DebugInfo/CXX/vtable-inheritance-simple.cpp#L31

Or, we can use default target triple as this:
https://github.com/llvm/llvm-project/blob/917f078adb3ffab4cf50aec3e13c977c092d7754/clang/test/Modules/ExtDebugInfo.cpp#L17-L20

However, since I can't verify non-x86 or non-linux target, I'd want to leave it on this PR.

File: clang/test/DebugInfo/CXX/vtable-template-instantiation.cpp (hunk 1 above)
Comment (kikairoya) [reply clang/test/DebugInfo/CXX/vtable-template-instantiation.cpp:16]: Yes, that's the cause of #149639, so we need to check that and to check if that doesn't happen for the COFF platform (in the next PR).

Oops, ```# for COFF targets:``` is odd here -- MinGW cases are not added yet.

**Review by jmorse — COMMENTED** [review]
This is looking good; my only remaining concern is about the use of `-O1` in a few places -- consulting the [testing guide](https://llvm.org/docs/TestingGuide.html) it really does encourage covering the smallest set of passes possible ("only run a minimal set of passes"). 

As you say, there are different outputs between -O0 and -O1, but is that because of a behaviour related to this feature, or is it just the normal behaviour of the compiler? If it's normal, we should remove the -O1 coverage. If there's something important (the presence / absence of the external symbol?) then we should narrow down what pass causes that behaviour and test that, rather than a full invocation of clang.

**Comment by jalopezg-git** [comment]
I am sorry for the delay in providing a review for this PR; it's on my TODO list for the next 24h though!

File: clang/test/DebugInfo/CXX/vtable-template-instantiation.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+//   * Its '_vtable$' is generated without associating to vtable
+//  # when non-optimized:
+//   * The vtable is generated with comdat only if non-optimized
+//   * Its '_vtable$' is generated regardless optimized or not
+// - Define explicitly instantiation:
+//   * The vtable is generated with comdat
+//   * Its '_vtable$' is generated
+// - Declare explicitly instantiation as extern:
+//   * The vtable is declared
+//   * Its '_vtable$' is generated only if optimized
```
Comment (jalopezg-git) [inline clang/test/DebugInfo/CXX/vtable-template-instantiation.cpp:17]: Let's improve the wording here a bit (for future readers).
```suggestion
// For the `CTemplate` templated class below, check the following cases:
// - Implicitly instantiate whole class by up-casting (`NOCAST` not defined):
//   * The vtable is generated with a COMDAT specifier
//   * Its '_vtable$' is generated
// - Implicitly instantiate member function only (`NOCAST` defined):
//  # when optimized:
//   * The vtable has no declaration
//   * Its '_vtable$' is generated, but not associated to vtable declaration
//  # when non-optimized:
//   * The vtable is generated with a COMDAT specifier
//   * Its '_vtable$' is generated regardless of optimization level
// - Define explicitly instantiation (`EXPLICIT` defined):
//   * The vtable is generated with a COMDAT specifier
//   * Its '_vtable$' is generated
// - Declare explicitly instantiation as `extern` (`EXTERN` defined):
//   * The vtable is declared
//   * Its '_vtable$' is generated only if optimized
```

File: clang/test/DebugInfo/CXX/vtable-template-instantiation.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+void CTemplate<T>::f() noexcept {}
+template <typename T>
+CTemplate<T>::~CTemplate() noexcept {}
+
+#ifdef EXPLICIT
+template struct CTemplate<void>;
+#endif
+#ifdef EXTERN
+extern template struct CTemplate<void>;
+#endif
```
Comment (jalopezg-git) [inline clang/test/DebugInfo/CXX/vtable-template-instantiation.cpp:38]: What about just
```suggestion
#if defined(EXPLICIT) || defined(EXTERN)
EXTERN template struct CTemplate<void>;
#endif
```
and specifying `-DEXTERN=extern` on the command line, where needed.

File: clang/test/DebugInfo/CXX/vtable-external.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+// - The definition of its destructor is visible:
+//   * The vtable is generated
+//   * Its '_vtable$' is generated
+// - Otherwise:
+//   * The vtable is generated
+//   * Its '_vtable$' is generated
+//
+// For CNoFnDef (member functions are declared only), we check in case of:
+//   * The vtable is NOT generated
+//   * Its '_vtable$' is generated only if optimized
```
Comment (jalopezg-git) [inline clang/test/DebugInfo/CXX/vtable-external.cpp:19]: Let's improve wording a bit also here, e.g.
```suggestion
// For the `CInlined` struct, where all member functions are inlined, we check the following cases:
// - If the definition of its destructor is visible:
//   * The vtable is generated with a COMDAT specifier
//   * Its '_vtable$' is generated
// - Otherwise:
//   * The vtable is declared
//   * Its '_vtable$' is NOT generated
//
// For `CNoInline` struct, where member functions are defined out-of-line, we check the following:
//   * The vtable is generated
//   * Its '_vtable$' is generated
//
// For `CNoFnDef` (member functions are declared only), we check that:
//   * The vtable is NOT generated
//   * Its '_vtable$' is generated only if optimizations are enabled
```

**Review by jalopezg-git — COMMENTED** [review]
Thank you!  I have added some minor comments / suggestions that add to other reviewer's comments :slightly_smiling_face:.

File: clang/test/DebugInfo/CXX/vtable-external.cpp (hunk 6 above)
Comment (kikairoya) [reply clang/test/DebugInfo/CXX/vtable-external.cpp]: I really appreciate you helping me improve my poor English!

File: clang/test/DebugInfo/CXX/vtable-template-instantiation.cpp (hunk 5 above)
Comment (kikairoya) [reply clang/test/DebugInfo/CXX/vtable-template-instantiation.cpp:36]: I think `-DEXPLICIT -DEXTERN=` (or just `-DEXTERN=`) and `-DEXTERN=extern` in the RUN lines don't make it clear what is supposed to happen, so I'd leave this.

File: clang/test/DebugInfo/CXX/vtable-inheritance-multiple.cpp [hunk 7]
```diff
@@ -30,23 +28,26 @@ struct CDerived : NSP_1::CBaseOne, NSP_2::CBaseTwo {
   int six() override { return 66; }
 };
 
+void use(void *, ...);
```
[the remark this answers is not in the document]
Comment (kikairoya) [reply clang/test/DebugInfo/CXX/vtable-inheritance-multiple.cpp]: The `use` anchors are no longer needed thanks to `-disable-llvm-passes` so they have been removed.

**Review by kikairoya — COMMENTED** [review]
I've updated to use `-disable-llvm-passes` and to address the reviews.
I've also made a change to the PR description.

**Review by jmorse — APPROVED** [review]
IMO this is ready to go; best to leave a day or two to see if anyone has further opinions to chip in.
